packages/components/src/autocomplete/types.ts

```diff
@@ -0,0 +1,212 @@
+/**
+ * WordPress dependencies
+ */
+import type { WPElement } from '@wordpress/element';
+/**
+ * External dependencies
+ */
+import type { MutableRefObject, ReactNode } from 'react';
+/**
+ * Internal dependencies
+ */
+import type { useAutocomplete } from '.';
+
+type GetOptions< TCompleterOption = any > = (
+	filterValue: AutocompleterUIProps[ 'filterValue' ]
+) => Array< TCompleterOption > | Promise< Array< TCompleterOption > >;
+
+export type OptionCompletion = {
+	action: 'insert-at-caret' | 'replace';
+	value: string | WPElement;
+};
+
+type OptionLabel = string | WPElement | Array< string | WPElement >;
+export type KeyedOption = {
+	key: string;
+	value: any;
+	label: OptionLabel;
+	keywords: Array< string >;
+	isDisabled: boolean;
+};
+
+export type WPCompleter< TCompleterOption = any > = {
+	/**
+	 * The name of the completer. Useful for identifying a specific completer to
+	 * be overridden via extensibility hooks.
+	 */
+	name: string;
+	/**
+	 * The string prefix that should trigger the completer. For example,
+	 * Gutenberg's block completer is triggered when the '/' character is entered.
+	 */
+	triggerPrefix: string;
+	/**
+	 * The raw options for completion. May be an array, a function that returns
+	 * an array, or a function that returns a promise for an array.
+	 * Options may be of any type or shape. The completer declares how those
+	 * options are rendered and what their completions should be when selected.
+	 */
+	options: Array< TCompleterOption > | GetOptions;
```

Comment:
Any reason why we're not passing `TCompleterOption` to `GetOptions` here?

```suggestion
	options: Array< TCompleterOption > | GetOptions< TCompleterOption >;
```

Reply:
No reason, this was an oversight on my part. It got addressed with the changes in 4bb27eecce, assuming we ultimately go that route.